src/core/core.controller.js

```diff
@@ -289,8 +289,14 @@ module.exports = function(Chart) {
 
 			helpers.each(me.data.datasets, function(dataset, datasetIndex) {
 				var meta = me.getDatasetMeta(datasetIndex);
+				var type = dataset.type || me.config.type;
 				if (!meta.type) {
-					meta.type = dataset.type || me.config.type;
+					meta.type = type;
+				} else if (meta.type !== type) {
+					meta.controller.destroy();
+					delete me.data.datasets[datasetIndex]._meta;
```

Comment:
meta are scoped per chart (`_meta[chart.id]`), each chart responsible to maintain their own meta data, so we should not delete all `_meta` here. Also, will be better to move that code in a new private method `destroyDatasetMeta`:

```javascript
/**
 * @private
 */
destroyDatasetMeta: function(datasetIndex) {
    var id = this.id;
    var dataset = this.data.datasets[datasetIndex];
    var meta = dataset._meta && dataset._meta[id];

    if (meta) {
        meta.controller.destroy();
        delete dataset._meta[id];
    }
}
```

This method can also be used in [`chart.destroy`](https://github.com/chartjs/Chart.js/blob/master/src/core/core.controller.js#L684-L688) to replace lines 684 to 688.